### ISSUE TITLE
Ideas for v4.0

### ISSUE THREAD

**Issue by jmejia8** [jmejia8/Metaheuristics.jl#12]
# Some ideas for v4

**Implement Flux-inspired algorithms**
Example:
```julia
# algorithm
GA = Chain(tournament(), SBX(p=0.9), PM(p=0.1), Evironmental(base=:nsga2))


# usage
P = InitialPopulation(N=100)
GA(P) # produces a new population

## optimize
for gen in 1:100
    P = GA(P)
end
```

**Comment by jmejia8** [comment]
As far as I know, it can be difficult to implement a viable relative converge criteria into metaheuristics (due to stochastic operations of metaheuristics). It would be nice for me to know some relative convergence criteria designed for metaheuristics...

Maybe, you can use the [`diff_check`](https://jmejia8.github.io/Metaheuristics.jl/stable/api/#Metaheuristics.diff_check) stopping criteria already implemented in `Metaheuristics.jl` as a relative convergence criteria:

**Example** (Metaheuristics v3.2.6):

```julia
julia> using Metaheuristics

julia> f, bounds, _ = Metaheuristics.TestProblems.sphere();

julia> options = Options(f_tol=1e-8,seed=1); # give tolerance

julia> optimize(f, bounds, ECA(;options))
+=========== RESULT ==========+
  iteration: 228
    minimum: 2.72864e-09
  minimizer: [-1.404483151684814e-5, -1.645523586390722e-5, 1.9944474758572303e-5, -8.336438266970471e-6, 1.3988146389886503e-5, -1.3177919277964173e-5, -3.0124814894731768e-6, 1.6324625110249603e-5, -1.8097438702973058e-5, -2.865158742490321e-5]
    f calls: 15947
 total time: 0.1724 s
stop reason: Small difference of objective function values.
+============================+
```

**Comment by fipelle** [comment]
I have to admit that I haven't explored the full potential of the package, so there may be cases where my suggestion does not apply. I am thinking mostly about cases in which the scale of the parameters is not uniform and you may want to have a relative convergence criterion such as ```median(abs(x-x_last_best)/abs(x_last_best+eps())```.

**Comment by jmejia8** [comment]
I got it. Could you share some bibliographic references on those relative convergence criteria (for metaheuristics)?

**Comment by fipelle** [comment]
Will do. For now https://www.sfu.ca/sasdoc/sashtml/iml/chap11/sect11.htm.

**Comment by mtzrene** [comment]
Hi, within the performance metrics, it would be interesting to add the multiplicative unitary epsilon. How difficult would it be to add it?

**Comment by jmejia8** [comment]
@mtzrene I didn't know such performance indicator. Could you provide more info about it? A possible integration within Metaheuristics can be carried out if a paper and/or author's code (Julia or not) are available.

**Comment by mtzrene** [comment]
@jmejia8 Of course, it is a performance criterion for multi-objective optimization. I looked for some code in another language to send you a reference, but was unsuccessful. However, it is mentioned in this article. https://ieeexplore.ieee.org/abstract/document/1197687

**Comment by jmejia8** [comment]
Thanks for the information, @mtzrene . I just read about the unary epsilon-indicator and it seems easy to implement. I have some questions about the *multiplicative* variant that we can discuss at issue #21

**Comment by pnovoa** [comment]
Hi Mejias, great project! Congratulations! It would be nice in future implementations to have a variant of "optimize" that performs the optimization process with restarts. This naive strategy could be very useful in multimodal problems and in those where we need to perform continuous optimization process (e.g. dynamic problems). It could be a new function (e.g. "optimize_with_restart") or some parameter inside "Options".

Another issue that I think would help a lot is to have a overloaded "copy" for State, because it is complex to transfer data between instances of that type, that is, because of the amount of parameters that State contains.

**Comment by jmejia8** [comment]
@pnovoa Thank you for your comments and suggestions. Could you share with me some bibliographic references on approaches using resstarts? Perhaps, a restarting procedure could be implemented as a wrapper:

```julia
method = Restart(GA(), every = 20 #= generations =#)
optimize(f, bounds, method)
```
Suggestions are welcome 😀 

Regarding the second issue, what would you expect the overloaded `copy` to do? Only copying the most important attributes?
